WIP: Firms update staging

mappum opened the pull request:
**EDIT:** This branch is now being treated as a staging branch for the firms update

----

*(Work in progress)*
Closes #272 
Closes #295

I'm setting up simple unit tests here to more easily finish the work on #271 without having to manually try each command on Reddit. This PR won't have 100% test coverage, but should save time by covering the core of the bot's behavior.

My plan is to mock out the `praw` comment API and `sqlachemy` session API enough to be able to run commands against the `CommentWorker`.

mappum committed:
Added Firm model

mappum committed:
Use re.fullmatch instead of re.search in order to allow commands that start with the same sequence of characters

mappum committed:
Implemented !createfirm command

mappum committed:
Implemented !leavefirm command

mappum committed:
Implemented !firm command

mappum committed:
Allow command arguments with casing

mappum committed:
Message fixes

mappum committed:
Fixed !leavefirm

mappum committed:
Show list of firm members in !firm reply

mappum committed:
Merge branch 'master' into firms

mappum committed:
Implemented !promote and !fire

mappum committed:
Implemented !joinfirm command

thecsw committed:
Added restore backup bash script

We have to restore databases easily and this just automates it

mappum committed:
Use python3 iteration

mappum committed:
Merge branch 'master' into firms

mappum committed:
Set up test package

mappum committed:
Added test script

mappum committed:
Only create virtual env and install deps if it does not exist

mappum committed:
Set up sqlite db for tests

mappum committed:
Skip comment instanceof praw check in the tests

mappum committed:
Added user init tests

mappum committed:
Isolate db state in each test so test case ordering doesn't matter

mappum committed:
Added a little more test coverage

mappum committed:
Mock praw models subclass str to match praw API

mappum committed:
Added !broke test cases

mappum committed:
Split test files

thecsw commented on docker-compose.yml:
```diff
@@ -48,3 +48,6 @@ services:
         restart: unless-stopped
+        command: ["mysqld", "--default-authentication-plugin=mysql_native_password"]
+        ports:
+            - 3306:3306
```
We don't want to expose our database! The credentials are open and database should be accessible only to the isolated network of containers

mappum replied:
Oops, didn't mean to commit that, was just exposing it locally for debugging. 😛

thecsw replied:
Okay!

thecsw commented on test/mock_praw.py:
```diff
@@ -0,0 +1,36 @@
+        return self.name
+
+class Submission(str):
+    def __new__(cls, id, *args):
```
l would recommend to change the `id` variable to something else as in python, `id` is a reserved keyword. It's really weird why they did that, idk. You can leave `id` is a var name in a method, so `self.id` or `comment.id` are okay to use but just using `id` alone (as a var) might be troublesome. You can rename it to `ID` or some C++ style `id_m`. Thanks

mappum committed:
Don't use 'id' for variable name

mappum committed:
Unexpose db port in container

mappum committed:
Lint

mappum committed:
Report test coverage

mappum committed:
Only report coverage if tests pass

mappum committed:
Fixed typo for insufficient funds message

mappum committed:
Added test coverage for non-firms commands

mappum committed:
Use integer timestamp value for sqlite

thecsw commented:
Hey! So I tested the PR, looks fantastic! I fetched the branch and when I ran the shell file `./test.sh`, it gave me the following error:

```
======================================================================
ERROR: balance (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: balance
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/loader.py", line 434, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.7/unittest/loader.py", line 375, in _get_module_from_name
    __import__(name)
  File "/home/thecsw/gits/memeinvestor_bot/test/balance.py", line 1, in <module>
    from test import Test
  File "/home/thecsw/gits/memeinvestor_bot/test/test.py", line 10, in <module>
    from comment_worker import CommentWorker
  File "src/comment_worker.py", line 12, in <module>
    import config
  File "src/config.py", line 19, in <module>
    USER_AGENT = os.environ['BOT_USER_AGENT']
  File "/usr/lib/python3.7/os.py", line 678, in __getitem__
    raise KeyError(key) from None
KeyError: 'BOT_USER_AGENT'
```

So, I tried to add all of the env vars to the shell script and export the values, just so `config.py` would ignore the errors if an env var is not found (you have to wove the export lines before the actual coverage test):

``` bash
sed -E '/[#]/d;/^$/d' .env.example | sed -E 's/.*/export &/g' >> test.sh
```

Then it ran! I got the logs, for some reason, some of them started with a dot:

```
INFO:root: -- testuser: !create
INFO:root: -- testuser: !invest 100
INFO:root: -- testuser: !invest 100
INFO:root: -- testuser: !active
.INFO:root: -- testuser: !create
INFO:root: -- testuser: !active
.INFO:root: -- testuser: !create
INFO:root: -- testuser: !balance
INFO:root: -- testuser: !balance
.INFO:root: -- testuser: !help
.INFO:root: -- testuser: !ignore
..INFO:root: -- testuser: !create
(43 lines were omitted for brevity)
INFO:root: -- autocreating
.INFO:root: -- testuser: !create
. (weird dot idk)
----------------------------------------------------------------------
Ran 26 tests in 9.319s
```

And then the final report came up:

```
OK
Name                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------
src/api.py                          141    141     44      0     0%
src/calculator.py                    93     93     18      0     0%
src/comment_worker.py               210     89     80      7    52%
src/config.py                        23      0      0      0   100%
src/describe-tables.py                8      8      0      0     0%
src/formula-graphs-for-paper.py      29     29      4      0     0%
src/formula-test.py                  44     44      8      0     0%
src/formula-visualizer.py           134    134     44      0     0%
src/formula.py                       34     34      6      0     0%
src/kill_handler.py                   7      7      0      0     0%
src/main.py                          69     69     24      0     0%
src/message.py                      118     18     12      1    82%
src/models.py                        42      1      2      1    95%
src/stopwatch.py                     12     12      0      0     0%
src/submitter.py                     73     73     24      0     0%
src/tops1.py                         27     27      8      0     0%
src/utils.py                         45     23     14      4    44%
-------------------------------------------------------------------
TOTAL                              1109    802    288     13    25%
```

So, the question is, did you have that error with exporting the variables? I know that you have `source .env` and it should work but it throws an error on my machine. I execute the script from the top directiory of the repo, so `.env` is accessed via `./.env`

**UPDATE** AH! I got the error. I don't know why it errors out on different machines, on my machine it's triggered due to a whitespace in the `.env` file.

Example:
```
.env: line 110: by: command not found
```

.env: line 110:
```
BOT_USER_AGENT=thecswbot by u/thecsw
```

The whitespace makes the script go nuts, to fix this, simply quote the whole thing as a UNIX tradition:

```
BOT_USER_AGENT="thecswbot by u/thecsw"
```

Very cool, Matt!

mappum committed:
Fix firm name length check

mappum committed:
Added test coverage for firm commands

thecsw commented:
I tested the last firms testing and got the following error about too many diffs:

```
======================================================================
FAIL: test_active (active.TestActive)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/thecsw/gits/memeinvestor_bot/test/active.py", line 30, in test_active
    Investment('post2', 100, 100, time.time())
AssertionError: '\nYo[114 chars]:00:07 remaining)\n\n[#2](https://www.reddit.c[72 chars]g)\n' != '\nYo[114 chars]:00:09 remaining)\n\n[#2](https://www.reddit.c[72 chars]g)\n'
Diff is 815 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 57 tests in 17.862s

FAILED (failures=1)
```

mappum commented:
Oh interesting, just looks like the tests are running slowly so the timestamps are different. I'll make so we can pass in a fixed timestamp instead of using `time.time()` in the tests.

thecsw committed:
Buffed out the formula.

r/MemeEconomy grew in sizes and it was time to update the formula so it would reflect those changes. I made the formula a bit tougher but still rewarding. It should decrease the exponential growth of some players. Thanks.

thecsw committed:
Added the formula plot.

thecsw committed:
Added a test variable.

This variable is needed for testing purposes. To add flairs, I had modify comment_worker.py directly. Our testing fails because comment_worker tries to initialize praw.Reddit instance that should not be created. To avoid that, I threw some if statements to check if the bot is in a testing environment.

thecsw committed:
Added test var in config file

thecsw committed:
Added the flairs system.

Flair system will append a flair to a user in a subreddit. If a user is in a subreddit, it will show that on his/her flair.

For example:
    Thecsw University of Memechester | CEO

thecsw committed:
Updated the flairs

thecsw committed:
Fixed the weird sqlalchemy.exc.ProgrammingError

thecsw committed:
Fixed the error with undefined variable.

thecsw committed:
Fixed the compile method for compile.

thecsw committed:
Fixed the describe-tables.py script

thecsw committed:
If user belongs to a firm, 15% is sent to the firm balance.

thecsw committed:
The formula is toughened.

thecsw committed:
Added more columns to the Firms table

 - Members has the number of firm members
 - Tax is the tax withholding percentage
 - Rank is the firm's rank (0-...)

thecsw committed:
Added firm columns manipulations

thecsw committed:
Added our patreon link

thecsw committed:
Merge pull request #289 from thecsw/flairs

Added the flairs system

thecsw committed:
Merge pull request #290 from thecsw/firm_tax

If user belongs to a firm, 15% is sent to the firm balance.

thecsw committed:
Merge pull request #292 from thecsw/patreon

Added our patreon link

thecsw committed:
Merge branch 'tests' into expand_firms

thecsw committed:
Added new columns to the Firm table.

firm.size -> shows the size of the firm (all members)
firm.execs -> number of executives in the firm (CEO not included)
firm.boost -> the profit boost that firm members git from their investments (1 is 0%, 1.1 is 10%), meant to be an upgrade
firm.tax -> the tax that all members of a firm should pay from their profits, needs to be regulated and limited, not 0% and not 100%, add validation steps in the code!
firm.rank -> firm's rank and the upgrade step, 1st lvl, 2nd lvl, and etc.

mappum committed:
Fix weird indentation

mappum committed:
Implemented !invite, !setprivate, and !setpublic commands

mappum committed:
Added tests for !invite

mappum committed:
Added leaderboard updater

mappum committed:
Added leaderboard updater

thecsw committed:
Added !template command

thecsw committed:
More personal now

thecsw committed:
Made python3 formatting style

thecsw committed:
Some formatting

thecsw committed:
Added error messages for the template command.

thecsw committed:
Added the version command.

Returns the date when teh bot was deployed

thecsw committed:
Formatting.

thecsw committed:
Updated the regex for security.

We wil capture only imgur links.

mappum committed:
Fixed numbers in leaderboard text

mappum committed:
Implemented createfirm cost, and updated tests

mappum committed:
Added firm membership limit

mappum committed:
Implemented executive limit

mappum committed:
Implemented !upgrade command

mappum committed:
Increased upgrade costs

mappum committed:
Deduct from firm balance on upgrade, and updated tests

thecsw committed:
Merge pull request #297 from thecsw/version

Added the version command.

mappum committed:
Merge branch 'tests' of github.com:MemeInvestor/memeinvestor_bot into tests

mappum committed:
Merge branch 'expand_firms' into tests

mappum committed:
Merge branch 'private-firms' into tests

mappum committed:
Merge branch 'template' into tests

mappum committed:
Fix !version command

mappum committed:
Added test for !version

mappum committed:
Merge branch 'firms-staging' into leaderboard

mappum committed:
Calculate net worth for leaderboard

mappum committed:
Format leaderboard as markdown table

mappum commented:
I merged the feature branches into here so we don't have to deal with the mess of branches later.

thecsw commented:
Awesome!

thecsw committed:
Hardcoded firm tax -> class variable

mappum committed:
Implemented payroll

mappum committed:
Merge branch 'payroll' into tests

mappum committed:
Merge remote-tracking branch 'thecsw/buff-formula' into tests

mappum committed:
Merge branch 'firm_tax_update' of https://github.com/thecsw/memeinvestor_bot into thecsw-firm_tax_update

mappum committed:
Merge remote-tracking branch 'thecsw/restore_database' into tests

thecsw committed:
Got rid of the boost attribute.

Stale. It is not used. Payroll will do the payouts

thecsw committed:
Changed zip release to the name of the package.

I am not entirely sure why we did that

thecsw committed:
Merge pull request #300 from thecsw/delete_boost

Got rid of the boost attribute.

thecsw committed:
Merge pull request #301 from thecsw/req_praw

Changed zip release to the name of the package.

thecsw reviewed (approved):
Approved, please do the honors! Don't forget to add your name to our list of contributors!
